File: src/utils/cacheService.js
/**
 * Cache Service
 * Provides in-memory caching for embeddings, query results, and other expensive operations.
 * 
 * Cache Types:
 * - Embedding cache: Stores text → embedding vector mappings
 * - Query cache: Stores complete query → response mappings
 * - Document cache: Stores document metadata/chunk counts
 */
const NodeCache = require('node-cache');
const crypto = require('crypto');

// ── Cache Configuration ──────────────────────────────────
const CACHE_CONFIG = {
  // Embedding cache - longer TTL since embeddings are deterministic
  embeddings: {
    ttl: parseInt(process.env.EMBEDDING_CACHE_TTL, 10) || 86400, // 24 hours
    maxKeys: parseInt(process.env.EMBEDDING_CACHE_MAX, 10) || 10000,
    checkperiod: 600 // Check for expired keys every 10 min
  },
  // Query cache - shorter TTL, can be invalidated by new docs
  queries: {
    ttl: parseInt(process.env.QUERY_CACHE_TTL, 10) || 3600, // 1 hour
    maxKeys: parseInt(process.env.QUERY_CACHE_MAX, 10) || 1000,
    checkperiod: 120
  },
  // Document metadata cache
  documents: {
    ttl: parseInt(process.env.DOC_CACHE_TTL, 10) || 1800, // 30 min
    maxKeys: 500,
    checkperiod: 300
  }
};

// ── Initialize Caches ────────────────────────────────────
const embeddingCache = new NodeCache({
  stdTTL: CACHE_CONFIG.embeddings.ttl,
  maxKeys: CACHE_CONFIG.embeddings.maxKeys,
  checkperiod: CACHE_CONFIG.embeddings.checkperiod,
  useClones: false // Embeddings are read-only, avoid cloning overhead
});

const queryCache = new NodeCache({
  stdTTL: CACHE_CONFIG.queries.ttl,
  maxKeys: CACHE_CONFIG.queries.maxKeys,
  checkperiod: CACHE_CONFIG.queries.checkperiod,
  useClones: true // Query results should be cloned
});

const documentCache = new NodeCache({
  stdTTL: CACHE_CONFIG.documents.ttl,
  maxKeys: CACHE_CONFIG.documents.maxKeys,
  checkperiod: CACHE_CONFIG.documents.checkperiod
});

// ── Helper Functions ─────────────────────────────────────

/**
 * Generate a hash key for a given input.
 * @param {string} input - Input text to hash
 * @returns {string} MD5 hash
 */
function hashKey(input) {
  return crypto.createHash('md5').update(input).digest('hex');
}

/**
 * Generate a cache key for a query with options.
 * @param {string} query - The query text
 * @param {Object} opts - Query options affecting the result
 * @returns {string} Cache key
 */
function buildQueryCacheKey(query, opts = {}) {
  const keyParts = [
    query,
    opts.topK || 5,
    opts.documentId || 'all',
    opts.rerank ? 'rerank' : 'norerank'
  ];
  return `q:${hashKey(keyParts.join('|'))}`;
}

// ── Embedding Cache ──────────────────────────────────────

/**
 * Get cached embedding for text.
 * @param {string} text - Input text
 * @returns {number[]|null} Cached embedding or null
 */
function getEmbedding(text) {
  const key = `emb:${hashKey(text)}`;
  return embeddingCache.get(key) || null;
}

/**
 * Cache an embedding for text.
 * @param {string} text - Input text
 * @param {number[]} embedding - Embedding vector
 * @param {number} ttl - Optional custom TTL in seconds
 */
function setEmbedding(text, embedding, ttl = null) {
  const key = `emb:${hashKey(text)}`;
  if (ttl) {
    embeddingCache.set(key, embedding, ttl);
  } else {
    embeddingCache.set(key, embedding);
  }
}

/**
 * Get multiple cached embeddings (returns array with nulls for misses).
 * @param {string[]} texts - Array of input texts
 * @returns {{ hits: Map<string, number[]>, misses: string[] }}
 */
function getEmbeddingsBatch(texts) {
  const hits = new Map();
  const misses = [];

  for (const text of texts) {
    const cached = getEmbedding(text);
    if (cached) {
      hits.set(text, cached);
    } else {
      misses.push(text);
    }
  }

  return { hits, misses };
}

/**
 * Cache multiple embeddings.
 * @param {Map<string, number[]>} embeddings - Map of text → embedding
 */
function setEmbeddingsBatch(embeddings) {
  for (const [text, embedding] of embeddings) {
    setEmbedding(text, embedding);
  }
}

// ── Query Cache ──────────────────────────────────────────

/**
 * Get cached query result.
 * @param {string} query - Query text
 * @param {Object} opts - Query options
 * @returns {Object|null} Cached result or null
 */
function getQueryResult(query, opts = {}) {
  const key = buildQueryCacheKey(query, opts);
  const cached = queryCache.get(key);
  if (cached) {
    // Mark as cache hit
    cached._fromCache = true;
    cached._cachedAt = cached._cachedAt || new Date().toISOString();
  }
  return cached || null;
}

/**
 * Cache a query result.
 * @param {string} query - Query text
 * @param {Object} opts - Query options
 * @param {Object} result - Query result to cache
 * @param {number} ttl - Optional custom TTL in seconds
 */
function setQueryResult(query, opts, result, ttl = null) {
  const key = buildQueryCacheKey(query, opts);
  const toCache = {
    ...result,
    _cachedAt: new Date().toISOString()
  };
  
  if (ttl) {
    queryCache.set(key, toCache, ttl);
  } else {
    queryCache.set(key, toCache);
  }
}

// ── Document Cache ───────────────────────────────────────

/**
 * Get cached document info.
 * @param {string} documentId
 * @returns {Object|null}
 */
function getDocumentInfo(documentId) {
  return documentCache.get(`doc:${documentId}`) || null;
}

/**
 * Cache document info.
 * @param {string} documentId
 * @param {Object} info
 */
function setDocumentInfo(documentId, info) {
  documentCache.set(`doc:${documentId}`, info);
}

// ── Cache Invalidation ───────────────────────────────────

/**
 * Invalidate all query cache entries (e.g., when new documents are added).
 */
function invalidateQueries() {
  queryCache.flushAll();
  console.log('[Cache] Query cache invalidated');
}

/**
 * Invalidate query cache entries for a specific document.
 * Note: This is approximate since keys are hashed.
 */
function invalidateDocumentQueries(documentId) {
  // Full invalidation for now - could be optimized with tagging
  queryCache.flushAll();
  documentCache.del(`doc:${documentId}`);
  console.log(`[Cache] Invalidated queries for document ${documentId}`);
}

/**
 * Clear all caches.
 */
function clearAll() {
  embeddingCache.flushAll();
  queryCache.flushAll();
  documentCache.flushAll();
  console.log('[Cache] All caches cleared');
}

// ── Statistics ───────────────────────────────────────────

/**
 * Get cache statistics.
 * @returns {Object} Stats for all caches
 */
function getStats() {
  return {
    embeddings: {
      keys: embeddingCache.keys().length,
      hits: embeddingCache.getStats().hits,
      misses: embeddingCache.getStats().misses,
      hitRate: calculateHitRate(embeddingCache.getStats())
    },
    queries: {
      keys: queryCache.keys().length,
      hits: queryCache.getStats().hits,
      misses: queryCache.getStats().misses,
      hitRate: calculateHitRate(queryCache.getStats())
    },
    documents: {
      keys: documentCache.keys().length,
      hits: documentCache.getStats().hits,
      misses: documentCache.getStats().misses
    }
  };
}

function calculateHitRate(stats) {
  const total = stats.hits + stats.misses;
  if (total === 0) return '0%';
  return `${((stats.hits / total) * 100).toFixed(1)}%`;
}

module.exports = {
  // Embedding cache
  getEmbedding,
  setEmbedding,
  getEmbeddingsBatch,
  setEmbeddingsBatch,
  
  // Query cache
  getQueryResult,
  setQueryResult,
  
  // Document cache
  getDocumentInfo,
  setDocumentInfo,
  
  // Invalidation
  invalidateQueries,
  invalidateDocumentQueries,
  clearAll,
  
  // Stats
  getStats,
  
  // Key helpers
  hashKey,
  buildQueryCacheKey
};

File: app/utils/cache_service.py
"""
Cache Service
In-memory TTL caching for embeddings, query results, and document metadata.

Cache tiers:
  - Embeddings : 24 h TTL  (deterministic — text always maps to same vector)
  - Queries    :  1 h TTL  (invalidated when the underlying documents change)
  - Documents  : 30 min TTL

Query/compare entries are indexed by the documents they touch, so deleting a
single document only evicts the entries that actually depend on it.
"""
import hashlib
import time
import os
from collections import OrderedDict
from typing import Any, Dict, Iterable, List, Optional, Set, Tuple

# ─── Configuration ─────────────────────────────────────────

EMBEDDING_TTL = int(os.getenv("EMBEDDING_CACHE_TTL", 86400))    # 24 hours
QUERY_TTL = int(os.getenv("QUERY_CACHE_TTL", 3600))             # 1 hour
DOC_TTL = int(os.getenv("DOC_CACHE_TTL", 1800))                 # 30 min
MAX_EMBEDDING_KEYS = int(os.getenv("EMBEDDING_CACHE_MAX", 10000))
MAX_QUERY_KEYS = int(os.getenv("QUERY_CACHE_MAX", 1000))

# Sentinel bucket for queries that search the whole corpus (no document filter).
_GLOBAL_SCOPE = "__all__"


class TTLCache:
    """LRU cache with per-key TTL and max-size eviction."""

    def __init__(self, default_ttl: int, max_keys: int = 10000):
        # OrderedDict gives us true LRU: most-recently-used lives at the end.
        self._store: "OrderedDict[str, Tuple[Any, float]]" = OrderedDict()
        self.default_ttl = default_ttl
        self.max_keys = max_keys

    def get(self, key: str) -> Optional[Any]:
        entry = self._store.get(key)
        if entry is None:
            return None
        value, expires_at = entry
        if time.time() > expires_at:
            del self._store[key]
            return None
        self._store.move_to_end(key)  # mark as recently used
        return value

    def set(self, key: str, value: Any, ttl: Optional[int] = None) -> None:
        if key in self._store:
            # Refresh in place so the entry moves to the MRU end below.
            del self._store[key]
        elif len(self._store) >= self.max_keys:
            self._evict_expired()
            while len(self._store) >= self.max_keys:
                self._store.popitem(last=False)  # drop least-recently-used
        self._store[key] = (value, time.time() + (ttl or self.default_ttl))
        self._store.move_to_end(key)

    def delete(self, key: str) -> None:
        self._store.pop(key, None)

    def flush(self) -> None:
        self._store.clear()

    def keys(self) -> List[str]:
        self._evict_expired()
        return list(self._store.keys())

    @property
    def size(self) -> int:
        # Cheap: count only unexpired entries without mutating the store.
        # (Expired entries are reaped lazily on get/set.)
        now = time.time()
        return sum(1 for _, exp in self._store.values() if now <= exp)

    def _evict_expired(self) -> None:
        now = time.time()
        expired = [k for k, (_, exp) in self._store.items() if now > exp]
        for k in expired:
            del self._store[k]


# ─── Cache Instances ───────────────────────────────────────

_embedding_cache = TTLCache(EMBEDDING_TTL, MAX_EMBEDDING_KEYS)
_query_cache = TTLCache(QUERY_TTL, MAX_QUERY_KEYS)
_document_cache = TTLCache(DOC_TTL, 500)

# document_id (or _GLOBAL_SCOPE) → set of query-cache keys that depend on it.
_doc_index: Dict[str, Set[str]] = {}


def _hash(text: str) -> str:
    return hashlib.md5(text.encode()).hexdigest()


def _build_key(prefix: str, subject: str, opts: dict) -> str:
    """
    Build a cache key from *every* option supplied.

    Options are serialized in sorted order, so adding a new retrieval knob
    automatically participates in the key instead of being silently dropped
    (which previously let a verify=True request hit a verify=False entry).
    """
    normalized = "|".join(f"{k}={opts[k]!r}" for k in sorted(opts))
    return f"{prefix}:{_hash(subject + '||' + normalized)}"


def _register(key: str, doc_ids: Iterable[Optional[str]]) -> None:
    """Index a cache key under each document it depends on."""
    scopes = {d or _GLOBAL_SCOPE for d in doc_ids} or {_GLOBAL_SCOPE}
    for scope in scopes:
        _doc_index.setdefault(scope, set()).add(key)


# ─── Embedding Cache ──────────────────────────────────────

def get_embedding(text: str):
    return _embedding_cache.get(f"emb:{_hash(text)}")

def set_embedding(text: str, embedding, ttl: Optional[int] = None):
    _embedding_cache.set(f"emb:{_hash(text)}", embedding, ttl)

def get_embeddings_batch(texts: List[str]):
    hits, misses = {}, []
    for t in texts:
        cached = get_embedding(t)
        if cached is not None:
            hits[t] = cached
        else:
            misses.append(t)
    return hits, misses

def set_embeddings_batch(mapping: dict):
    for text, emb in mapping.items():
        set_embedding(text, emb)


# ─── Query Cache ──────────────────────────────────────────

def get_query_result(query: str, opts: dict = {}):
    return _query_cache.get(_build_key("q", query, opts))

def set_query_result(query: str, opts: dict, result: dict, ttl: Optional[int] = None):
    key = _build_key("q", query, opts)
    _query_cache.set(key, result, ttl)
    # A query with no document_id searches everything, so it lands in the
    # global bucket and is evicted whenever any document changes.
    _register(key, [opts.get("document_id")])


# ─── Compare Cache ────────────────────────────────────────

def get_compare_result(topic: str, opts: dict = {}):
    return _query_cache.get(_build_key("c", topic, opts))

def set_compare_result(topic: str, opts: dict, result: dict, ttl: Optional[int] = None):
    key = _build_key("c", topic, opts)
    _query_cache.set(key, result, ttl)
    _register(key, opts.get("document_ids") or [])


# ─── Invalidation ─────────────────────────────────────────

def invalidate_queries():
    """Drop every cached query/compare result (used when the corpus grows)."""
    _query_cache.flush()
    _doc_index.clear()

def invalidate_document_queries(document_id: str):
    """
    Evict only the entries that depend on *document_id*.

    That means results scoped to this document, compares involving it, and
    whole-corpus queries (which may have retrieved its chunks). Results scoped
    to other specific documents are left intact.
    """
    keys = _doc_index.pop(document_id, set()) | _doc_index.get(_GLOBAL_SCOPE, set())
    for key in keys:
        _query_cache.delete(key)
    _doc_index.pop(_GLOBAL_SCOPE, None)
    # Drop the evicted keys from any remaining document buckets.
    for scope in list(_doc_index):
        _doc_index[scope] -= keys
        if not _doc_index[scope]:
            del _doc_index[scope]


# ─── Document Cache ───────────────────────────────────────

def get_document(document_id: str):
    return _document_cache.get(f"doc:{document_id}")

def set_document(document_id: str, data: dict):
    _document_cache.set(f"doc:{document_id}", data)


# ─── Stats ────────────────────────────────────────────────

def get_stats() -> dict:
    return {
        "embeddings": _embedding_cache.size,
        "queries": _query_cache.size,
        "documents": _document_cache.size,
    }

def flush_all():
    _embedding_cache.flush()
    _query_cache.flush()
    _document_cache.flush()
    _doc_index.clear()
